### DISCUSSION TITLE
Not able to configure Apple provider

### DISCUSSION THREAD

**Discussion by pffigueiredo** [strapi/strapi#6455] [Legacy Topics]
I was trying to add a custom provider to my strapi backend, in this case **Apple**, and as such, I was following the [docs](https://strapi.io/documentation/v3.x/plugins/users-permissions.html#configure-your-oauth-generic-information).

I added the missing files:

- extensions/users-permissions/services/Providers.js
- extensions/users-permissions/config/functions/bootstrap.js
- extensions/users-permissions/admin/src/components/PopUpForm/index.js
- extensions/users-permissions/admin/src/translations/en.json

as I didn't have any of this in my current project and added the following configuration to Providers.js file:

    case "apple": {
      
    const apple = purest({
        provider: "apple",
        config: {
          apple: {
            "https://appleid.apple.com": {
              __domain: {
                auth: {
                  qs: {
                    access_token: "[0]",
                  },
                },
              },
              "{endpoint}": {
                __path: {
                  alias: "__default",
                },
              },
            },
          },
        },
      });

      apple
        .query()
        .get("auth/authorize")
        .auth(access_token)
        .request((err, res, body) => {
          if (err) {
            callback(err);
          } else {
            callback(null, {
              username: body.id_token.email,
              email: body.id_token.email,
            });
          }
        });

And the following in bootstrap.js:

       const grantConfig = {
    email: {
      enabled: true,
      icon: "envelope",
    },
    apple: {
      enabled: true,
      icon: "apple",
      key: "",
      secret: "",
      callback: `${process.env.SERVER_URL}/auth/apple/callback`,
      scope: ["identify", "email"],
    }, 
    ...

And all I get when going to localhost:[SERVER_PORT]/connect/apple is: this url http://localhost:3000/connect/apple?error=Grant%3A%20missing%20or%20misconfigured%20provider

Which is the correct redirect url "locahost:3000/connect/apple/", but with an error saying that the provider is either missing or misconfigured.

I've also tried to replace all the code inside the apple provider, with the discord one, as this should be well configured, as it is strapi's code. But the error continues as if, the Providers.js file isn't even being read from.

Do you have any tips on how can I debug this further? Any help is appreciated.

**Comment by madarche** [comment]
See #7206

**Comment by nitewulf** [comment]
I think what the mentioned #7206 is getting at but in my opinion not explicit about is that you need to modify addition configuration files:

/node_modules/grant/config/oauth.json
/node_modules/grant/config/profile.json

and add entries there for your custom provider

I also had to modify /config/server.js so that the redirect/callback URLs are generated properly for the Providers.

need to add: url: 'http://yourdomain:1337'
